### PR TITLE
Enhance mock test for dynamic buffer manager for port removing and qos reload flows

### PR DESCRIPTION
**Pull request by stephenxs**
**What I did**
1. Enhance the mock test of the dynamic buffer manager in port remove and config qos clear flow and fix bugs during mock test implementation
2. Implement mock method `ProduceStateTable::del`

Signed-off-by: Stephen Sun <stephens@nvidia.com>

**Why I did it**

**How I verified it**
Run regression test, mock test, vs test, and manual test.

**Details if related**
1. Support mock test for dynamic buffer manager
   - `config qos clear` and reclaiming buffer
   - Remove port
3. Handle port remove/create flow
   - Cache cable length for a port
   - Try reclaiming unused buffer when maximum buffer parameters are received for a port whose state is `ADMIN_DOWN` and `m_bufferCompletelyInitialized` is `true`
4. Handle `config qos clear`
   - If all buffer pools are removed when `m_bufferPoolReady` is true, remove all zero pools and profiles.
   - Reload zero profiles and pools if they have not been loaded when reclaiming buffer

### CODE REVIEW

**Comment by stephenxs** [comment]
Most of the failures are warmreboot-related. None of the failures is related to the change in this PR. As this is still a draft and depends on other PR which has not yet been created, I will not retrigger it for now.

**Comment by liat-grozovik** [comment]
@stephenxs  well done on adding more UT to swss

**Comment by liat-grozovik** [comment]
@neethajohn could you please help to review?

**Comment by liat-grozovik** [comment]
@stephenxs  please handle conflicts so we can move on with the merge

**Comment by stephenxs** [comment]
> @stephenxs please handle conflicts so we can move on with the merge

sure. But this also depends on another PR #2234

**Comment by stephenxs** [comment]
vs test failed due to `##[error]Microsoft will start blocking usage of non-exempt self-hosted Azure DevOps pools starting April‘22.`.

**Comment by liat-grozovik** [comment]
@neethajohn , @prsunny kindly reminder to review this PR

File: cfgmgr/buffermgrdyn.cpp [hunk 1]
```diff
@@ -2304,6 +2326,20 @@ task_process_status BufferMgrDynamic::handlePortTable(KeyOpFieldsValuesTuple &tu
...
         }
     }
+    else if (op == DEL_COMMAND)
+    {
+        if (m_portPgLookup.find(port) != m_portPgLookup.end()
+            || m_portQueueLookup.find(port) != m_portQueueLookup.end()
+            || m_portProfileListLookups[BUFFER_INGRESS].find(port) != m_portProfileListLookups[BUFFER_INGRESS].end()
+            || m_portProfileListLookups[BUFFER_EGRESS].find(port) != m_portProfileListLookups[BUFFER_EGRESS].end())
+        {
+            SWSS_LOG_NOTICE("Port %s can be removed before buffer items have been removed", port.c_str());
```
Comment (prsunny) [inline cfgmgr/buffermgrdyn.cpp:2336]: Please change log level to INFO for retry logic.

File: cfgmgr/buffermgrdyn.cpp [hunk 2]
```diff
@@ -2634,6 +2690,10 @@ void BufferMgrDynamic::handleSetSingleBufferObjectOnAdminDownPort(buffer_directi
     {
         if (idsToZero.empty())
         {
+            // Happens only after "config qos reload"
+            if (!m_zeroProfilesLoaded)
```
Comment (prsunny) [inline cfgmgr/buffermgrdyn.cpp:2694]: please add` {} `

File: cfgmgr/buffermgrdyn.cpp [hunk 3]
```diff
@@ -3462,6 +3544,16 @@ void BufferMgrDynamic::handlePendingBufferObjects()
     }
 }
 
+void BufferMgrDynamic::cleanUpItemsForReclaimingBuffer(const string &port)
+{
+    // Clean up zero buffers when the buffer pools or a port has been removed
+    if (!m_bufferObjectIdsToZero[BUFFER_PG].empty())
```
Comment (prsunny) [inline cfgmgr/buffermgrdyn.cpp:3550]: please add `{}`

**Review by prsunny — COMMENTED** [review]
@neethajohn to review on functionality.

File: cfgmgr/buffermgrdyn.cpp (hunk 2 above)
Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:2694]: ok. will fix all.

Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:2706]: all fixed.

**Comment by stephenxs** [comment]
Failure is caused by timeout.

**Comment by liat-grozovik** [comment]
@neethajohn  could you please help to review and signoff?

**Review by prsunny — APPROVED** [review]
lgtm, wait for @neethajohn's approval

**Review by neethajohn — APPROVED** [review]

**Review by keboliu — APPROVED** [review]

**Comment by yxieca** [comment]
@stephenxs  this change cannot be cherry-picked into 202205 branch cleanly. please raise PR.

**Comment by stephenxs** [comment]
@yxieca thanks for update.
I forgot to mention dependency.
we need to cherry-pick https://github.com/azure/sonic-swss/pull/2234 first and then this one can be cherry-picked cleanly.

**Comment by yxieca** [comment]
@stephenxs thanks. done.

**Pull request merged**